node/src/components/storage.rs

```diff
@@ -1422,38 +1695,302 @@ impl Storage {
+        while current_digest != hash::SENTINEL1 && !live_digests[0].contains(&current_digest) {
+            live_digests[0].insert(current_digest);
+            let [key_to_part_db, merkle_proof_of_rest]: [Digest; 2] =
+                match txn.get_value(*block_body_v2_db, &current_digest)? {
+                    Some(slice) => slice,
+                    None => {
+                        return Err(Error::CouldNotFindBlockBodyPart {
+                            block_hash: header.hash(),
+                            merkle_linked_list_node_hash: current_digest,
+                        })
+                    }
+                };
+            if live_digests_index < live_digests.len() {
+                live_digests[live_digests_index].insert(key_to_part_db);
+            } else {
+                return Err(Error::UnexpectedBlockBodyPart {
+                    block_body_hash: *body_hash,
+                    part_hash: key_to_part_db,
+                });
+            }
+            live_digests_index += 1;
+            current_digest = merkle_proof_of_rest;
+        }
+    }
 
-        if should_check_integrity {
-            let body: BlockBody = lmdb_ext::deserialize(raw_val)?;
-            assert_eq!(
-                raw_key,
-                body.hash().as_ref(),
-                "found corrupt block body in database"
-            );
+    let databases_to_clean: [(&Database, &str); BlockBody::PARTS_COUNT + 1] = [
+        (block_body_v2_db, "deleting v2 block body part"),
+        (deploy_hashes_db, "deleting v2 deploy hashes entry"),
+        (transfer_hashes_db, "deleting v2 transfer hashes entry"),
+        (proposer_db, "deleting v2 proposer entry"),
+    ];
+
+    // Clean dead entries from all the databases
+    for (index, (database, info_text)) in databases_to_clean.iter().enumerate() {
+        let mut cursor = txn.open_rw_cursor(**database)?;
+        for (raw_key, _raw_val) in cursor.iter() {
+            let key = Digest::try_from(raw_key)
+                .map_err(|err| LmdbExtError::DataCorrupted(Box::new(err)))?;
+            if !live_digests[index].contains(&key) {
+                info!(?raw_key, info_text);
+                cursor.del(WriteFlags::empty())?;
+            }
+        }
+        drop(cursor);
+    }
+
+    Ok(())
+}
+
+/// Purges stale entries from the (Merklized) block body database, and checks the integrity of the
+/// remainder if `should_check_integrity` is true.
+#[allow(clippy::too_many_arguments)]
+fn initialize_block_body_v2_db(
+    env: &Environment,
+    block_header_db: &Database,
+    block_body_v2_db: &Database,
+    deploy_hashes_db: &Database,
+    transfer_hashes_db: &Database,
+    proposer_db: &Database,
+    any_v2_block_deleted: bool,
+    should_check_integrity: bool,
+) -> Result<(), Error> {
+    info!("initializing v2 block body database");
+
+    let txn = env.begin_rw_txn()?;
+
+    let block_body_hash_to_header_map = if any_v2_block_deleted || should_check_integrity {
```

Comment:
It looks like we're only using this if `should_check_integrity`. Maybe this can just be moved into the `if` branch?

Reply:
Good point - I think it got left in this branch of execution accidentally after garbage collection was disabled.